### ISSUE TITLE
Keyboards listed on iPad when there is no iPad definition available

### ISSUE THREAD

**Issue by snomos** [divvun/divvun-dev-keyboard#4]
On the iPad, when a keyboard layout only has an iPhone layout definition, it is still listed and available, but returns a US keyboard (Livonian in this case):

![image](https://github.com/user-attachments/assets/c15ea74f-e7ca-4c97-848d-ee76f4dbdfab)

In General > keyboards:

![image](https://github.com/user-attachments/assets/c4b3e1e2-3184-48cf-9e72-4ef74cae17c7)

In the keyboard selection menu:

https://github.com/user-attachments/assets/bc1bb498-864c-46a8-8f1b-5b0b5a727043

It would be better to hide the keyboard completely when no suitable definition for the device is available. As it is now it is very confusing for users.

I will nevertheless release this version, as it fixes the core issue on the iPhone, the iPad fixes can be released separately.

**Comment by dylanhand** [comment]
I've been looking into this one today. The expected solution, one might think, would be to disable iPad as a supported destination, here:

![CleanShot 2024-11-19 at 13 10 57@2x](https://github.com/user-attachments/assets/d095f470-5c50-4aa8-8d82-1a5055adb51d)

In testing however, this has no effect, possibly because iPhone apps are able to run on iPad. To test sanity I even made a separate project with keyboard extension and disabled everything except iPhone as a Supported Destination on each target including the hosting app. The only difference was that the hosting app appeared as an iPhone app in iPad mode (with an option to zoom in or rotate in bottom right corner), but the keyboard was still listed on iPad and could be enabled.

Given this limitation, an alternative solution would be to show a message on iPad where the keyboard should appear, something like:

`"The <insert language> keyboard is currently only available on iPhone".`

That way the user gets some indication of why it isn't working instead of doing nothing and defaulting to the US keyboard.

What do you think @snomos?

**Comment by snomos** [comment]
Yes. And a note to contact the developer to request iPad support, with an email address or something. That way we would get an indication about which layouts to prioritise, if needed.

**Comment by dylanhand** [comment]
@snomos for the feature that allows a user to notify us if they'd like to see a particular keyboard supported on iPad, I'm imagining a button that pre-fills an email with the keyboard language/locale, device type, and an email address to which these requests should be sent. Something like:

```
To: keyboard-requests@divvun.org
Subject: Request for keyboard-liv on iPad 9inch
Body: <User adds any additional notes here>
```

It appears it's not possible to pre-fill an email in Mail.app from an app extension, but it should be possible by first launching the HostingApp and doing it from there.

Checking with you before I go further in case you had something else in mind:)

**Comment by snomos** [comment]
Sounds good. To make the extra step via the app more useful, the email text should be presented to the user, and possibly in an editable form (or is the editing done in the Mail app?). The recipient email address should be configurable in the keyboard app repo, by default it should be `feedback@divvun.no`.

You could also have another button or link that instead take the user to the relevant GitHub repository/issues, with some pre-filled text. That should also be possible, along these lines:

https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/creating-an-issue#creating-an-issue-from-a-url-query

`https://github.com/giellalt/keyboard-liv/issues/new?labels=enhancement&title=Request+for+keyboard-liv+on+iPad+9inch`

This is obviously only useful for users with GitHub accounts, but for such users it might be handier.

**Comment by dylanhand** [comment]
I can make it possible to submit a request both via email and GitHub. And for email, it should be possible for us to bring up a Mail.app compose window (without leaving HostingApp) that is pre-filled and editable by the user.

I'll also look into how to make the email address configurable. It appears that [`.kbdgen` bundles already contain an email address](https://github.com/giellalt/keyboard-fit/blob/main/fit.kbdgen/project.yaml), but I'll need to dig into kbdgen a bit to see how/if it's being used.

**Comment by snomos** [comment]
Sounds good, all of the above.

**Comment by dylanhand** [comment]
New version available now in TestFlight that has most everything done, full list here: https://github.com/divvun/giellakbd-ios/pull/237

**Comment by dylanhand** [comment]
This task should now be complete, including getting the contact email address dynamically from each keyboard's `project.yaml` inside the `.kbdgen` bundle.

Here's a screenshot from Divvun Dev Keyboards 1.1.2 (247) in TestFlight using the contact email address temporarily changed in keyboard-liv in this commit: https://github.com/giellalt/keyboard-liv/commit/72058bd45604616453463e2b62e2e20dc253cce1

![Image](https://github.com/user-attachments/assets/c9b0af59-eb3e-42b9-b03d-460d176dc1be)

Note that `keyboard-liv` seemed to be the only keyboard without a layout on iPad in the Divvun Dev Keyboards app.

Feel free to update the email addresses in each keyboard's `project.yaml` and close this if you agree it's complete @snomos :)

**Comment by snomos** [comment]
Tested and released. Closed.
